tools-v2/pkg/cli/command/curvebs/peer/utils/copyset.go

```diff
@@ -0,0 +1,77 @@
+/*
+ *  Copyright (c) 2022 NetEase Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * Project: tools-v2
+ * Created Date: 2022-11-30
+ * Author: zls1129@gmail.com
+ */
+
+package utils
+
+import (
+	"context"
+	"fmt"
+
+	"google.golang.org/grpc"
+
+	cmderror "github.com/opencurve/curve/tools-v2/internal/error"
+	basecmd "github.com/opencurve/curve/tools-v2/pkg/cli/command"
+	common "github.com/opencurve/curve/tools-v2/proto/proto/common"
+	"github.com/opencurve/curve/tools-v2/proto/proto/copyset"
+)
+
+// DeleteBrokenCopysetRPC the rpc client for the rpc function DeleteBrokenCopyset
+type DeleteBrokenCopysetRPC struct {
+	Info    *basecmd.Rpc
+	Request *copyset.CopysetRequest
+	Cli     copyset.CopysetServiceClient
+}
+
+var _ basecmd.RpcFunc = (*DeleteBrokenCopysetRPC)(nil) // check interface
+
+// NewRpcClient ...
+func (ufRp *DeleteBrokenCopysetRPC) NewRpcClient(cc grpc.ClientConnInterface) {
+	ufRp.Cli = copyset.NewCopysetServiceClient(cc)
+}
+
+// Stub_Func ...
+func (ufRp *DeleteBrokenCopysetRPC) Stub_Func(ctx context.Context) (interface{}, error) {
+	return ufRp.Cli.DeleteBrokenCopyset(ctx, ufRp.Request)
+}
+
+// DeleteBrokenCopyset delete broken copyset.
+func DeleteBrokenCopyset(logicalPoolID, copysetID uint32, peer *common.Peer, opts Options) error {
+	rpcCli := &DeleteBrokenCopysetRPC{
+		Request: &copyset.CopysetRequest{
+			LogicPoolId: &logicalPoolID,
+			CopysetId:   &copysetID,
+		},
+		Info: basecmd.NewRpc([]string{peer.GetAddress()}, opts.Timeout, opts.RetryTimes, "DeleteBrokenCopyset"),
+	}
+	response, errCmd := basecmd.GetRpcResponse(rpcCli.Info, rpcCli)
+	if errCmd.TypeCode() != cmderror.CODE_SUCCESS {
+		return fmt.Errorf("failed to delete broken copyset %s error : %s", peer.GetAddress(), errCmd.Message)
+	}
+	resp, ok := response.(*copyset.CopysetResponse)
+	if !ok {
+		return fmt.Errorf("error type interface when acquire delete broken copyset info %s", peer.GetAddress())
+	}
+	if resp.GetStatus() != copyset.COPYSET_OP_STATUS_COPYSET_OP_STATUS_SUCCESS {
+		return fmt.Errorf("delete broken failed, error code %v", resp.GetStatus())
+	}
+	return nil
```

Comment:
the same as  `query leader `, should we add  it ?

Reply:
> the same as `query leader `, should we add it ?

It may not be a subcommand of the corresponding command, but please organize the file according to the corresponding structure.

Reply:
It is not a command, just the function be invoked during the remove peer process. I move it to the peer dir.  https://github.com/opencurve/curve/blob/5e49aed29c6cbf75f046373ecfd17ec0d3fbe0c5/src/tools/curve_cli.cpp#L137

Reply:
> It is not a command, just the function be invoked during the remove peer process. I move it to the peer dir.
> 
> https://github.com/opencurve/curve/blob/5e49aed29c6cbf75f046373ecfd17ec0d3fbe0c5/src/tools/curve_cli.cpp#L137

What I mean is that this command can be used as a `curve bs query leader`, even if it is not needed at present, it can be added in query.go if it is needed in the future.

Reply:
the result of the old command `curve_ops_tool check-copyset --logicalPoolId=1 --copysetId=29 --detail` can query the details of copyset which incloud the leader info. We can implement  the corresponding command in new tool other than use `query leader` command.

the result of  `check-copyset`:
```
[4294967325]
peer_id: 127.0.0.1:8201:0
state: LEADER
readonly: 0
term: 11
conf_index: 8
peers: 127.0.0.1:8200:0 127.0.0.1:8201:0
changing_conf: NO    stage: STAGE_NONE
election_timer: timeout(1000ms) STOPPED
vote_timer: timeout(1000ms) STOPPED
stepdown_timer: timeout(1000ms) SCHEDULING(in 393ms)
snapshot_timer: timeout(1800000ms) SCHEDULING(in 977625ms)
storage: [9, 8]
disk_index: 8
known_applied_index: 8
last_log_id: (index=8,term=11)
state_machine: Idle
last_committed_index: 8
last_snapshot_index: 8
last_snapshot_term: 11
snapshot_status: IDLE
replicator_16840566767634@127.0.0.1:8200:0: next_index=9  flying_append_entries_size=0 idle hc=4226643 ac=7 ic=0

```